O2/ML/prepare_samples.py

```diff
@@ -20,6 +21,50 @@
                "XicToPKPi": 4}
 
 
+def do_dca_smearing(df, nProng=None):
+    """
+    Method to do the DCA smearing for 2 prongs and 3 prongs
+
+    Parameters
+    -----------------
+    - df: pandas dataframe containing all candidates with fFlagOrigin column
+    - nProng: option to 2 prongs or 3 prongs
+
+    Outputs
+    -----------------
+    - df: New dataframe with the smeared DCA columns
+    """
+    
+    print("Start to do the smearing")
+    # Open the input files
+    file_names = ["sigmaDcaXY_LHC22q_pass2_LHCC.root", "sigmaDcaZ_LHC22q_pass2_LHCC.root"]
+    input_files = [TFile.Open(name) for name in file_names]
+    
+    # Extract DCA resolution histograms
+    dca_reso = {}
+    for i, par in enumerate(["XY", "Z"]):
+        gDca = input_files[i].Get("can")
+        dca_reso[par] = gDca.GetPrimitive(f"tge_DCA_res_withPVrefit_all")
+    
+    # Add smeared DCA columns to the dataframe
+    smear_cols = ["XY1", "XY2", "Z1", "Z2"]
+    if nProng == 3:
+        smear_cols.extend(["XY3", "Z3"])
+    for col in smear_cols:
+        dca_col = f"fDCAPrim{col}"
+        pt_col = f"fPT{col[-1]}"
+        df[f"{dca_col}_SMEAR"] = [
+            gRandom.Gaus(dca, dca_reso[col[:-1]].Eval(pt) * 1e-4)
```

Comment:
Since our impact parameters are already smeared, applying a smearing that corresponds to the dca in data would result in a too large resolution in the end.

You should instead applying a smearing of an "effective" sigma which is 
```
sigma_smear = sqrt(sigma(DCA)_data^2 - sigma(DCA)_MC^2)
```

 to obtain in the end a smearing that is equal to the one we have in data.

Reply:
Hi @fgrosa, Thanks a lot! I ask two very stupid question:

- We do the smearing ("effective" sigma ) on the data because the data is MB MC production for the moment, right? 
- Then we do smearing with the data resolution for the ccbar and bbbar production, right?

Reply:
Ciao @zhangbiao-phy!
- we use the "effective" sigma because our impact parameter distributions are already "smeared" given that they are reconstructed, they just have a smaller resolution
- the impact-parameter resolution does not depend on the MC production (it is the same for general purpose and HF enriched)